internal/server/server.go

```diff
@@ -0,0 +1,240 @@
+		return
+	}
+
+	c.Header("Content-Type", "text/html")
+	sessionCookie := http.Cookie{
+		Name:     "session_token",
+		Value:    session.SessionToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+	}
+
+	// Gin has a slightly weirder interface for cookies for some reason
+	http.SetCookie(c.Writer, &sessionCookie)
+	c.HTML(http.StatusOK, "index.html", gin.H{"csrf": csrfToken})
+}
+
+func (env *Env) login(c *gin.Context) {
+	now := time.Now()
+	csrfToken := c.Request.Header.Get("CSRF")
+
+	sessionToken, err := c.Cookie("session_token")
+	// should look into deduplicating this
+	// should also look into a logging library w/ levels (debug, info, warning, etc)
+	if err != nil {
+		log.Println("Missing session token")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	session, err := env.db.FetchSession(sessionToken)
+	if err != nil {
+		log.Printf("Error fetching session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if session.Expired(now) {
+		log.Println("Session expired")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if !crypto.IsSessionOwner(session, csrfToken) {
+		log.Println("Not owner of session")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if session.Authenticated() {
+		log.Println("Already logged in")
+		c.AbortWithStatus(http.StatusOK)
+		return
+	}
+
+	email, password, ok := c.Request.BasicAuth()
+	if !ok {
+		log.Println("Missing basic auth")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	account, err := env.db.FetchAccount(email)
+	if err != nil {
+		log.Println("Missing account")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	isCorrectPassword, err := crypto.IsCorrectPassword(account, password)
+	if err != nil {
+		log.Println("Salt stored in database is incorrectly encoded")
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	if !isCorrectPassword {
+		log.Println("Bad password")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	// both session tuples should only be accessed by this user, so we won't bother
+	// with a transaction
+	authenticatedSession, err := env.db.CreateSession(account.AccountId, csrfToken, now)
+	if err != nil {
+		log.Printf("Error creating session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	err = env.db.DeleteSession(sessionToken)
+	if err != nil {
+		log.Printf("Error deleting old session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	sessionCookie := http.Cookie{
+		Name:     "session_token",
+		Value:    authenticatedSession.SessionToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+	}
+
+	http.SetCookie(c.Writer, &sessionCookie)
+	c.AbortWithStatus(http.StatusOK)
+}
+
+func (env *Env) logout(c *gin.Context) {
+	// TODO: move logged in check to its own function
```

Comment:
Yeah, I think if you move this logic for checking csrf token, retrieving and checking the sesion etc, that would make it a bit cleaner.